### PR TITLE
[jit] list add insert and remove

### PR DESCRIPTION
**Pull request by TheCodez**
See https://github.com/pytorch/pytorch/issues/16662

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -1051,6 +1052,91 @@ int listClear(Stack& stack) {
...
+
+template <>
+int listRemove<Shared<GenericList>, IValue>(Stack& stack) {
+  Shared<GenericList> list;
+  IValue elem;
+  pop(stack, list, elem);
+
+  auto& elements = list->elements();
+  auto pos = std::find_if(elements.begin(), elements.end(), [elem](const IValue& b) {
+    return b.isAliasOf(elem);
```
Comment (TheCodez) [inline torch/csrc/jit/register_prim_ops.cpp]: @eellison is this the correct way to check for equality of IValues?

Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: Hi, 

Python list remove checks each element of the list for value equality to the input. IsAliasOf checks reference equality. Basically the difference between using `a == b` and `a is b`. You probably are going to want to do some refactoring and use our existing implementations of the equality operator "aten::eq." Thanks for the PR, looks good!!

Comment (TheCodez) [reply torch/csrc/jit/register_prim_ops.cpp]: In which case would the `IValue` case be called?

Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: IsAliasOf is in the GenericList implementation, which means all the lists we don't have existing specializations for - everything but tensors, ints, floats, and bools. For example, a list of int lists. Pretty sure [[.1, .2], [.5], [1.0]].remove([.5]) would remove an element in python but would not work currently in torchscript. 

FYI: a [writeup](https://github.com/pytorch/pytorch/blob/f42f7c95b301ba02df04f722393e5b5da4d41718/torch/csrc/jit/README.md#ivalue) on ivalues & other technicals

Comment (TheCodez) [reply torch/csrc/jit/register_prim_ops.cpp]: Thanks for the help, didn't think of that case 👍 The document looks really helpful btw!
I couldn't really come up with a solution on how to use aten::eq, so my idea currently is to move the remove op out of CREATE_MUTABLE_LIST_OPS and move it to its own CREATE_REMOVE_OP, which is defined for TensorList and GenericList.
And then do something along the lines of [ListConstruct](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/register_prim_ops.cpp#L786-L793).
Maybe I'm overcomplicating things but I can't think of a better solution atm.

So any more help would be much appreciated.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -1051,6 +1052,91 @@ int listClear(Stack& stack) {
...
+  int64_t idx;
+  TElement elem;
+  pop(stack, list, idx, elem);
+
+  auto& elements = list->elements();
+  const int64_t list_size = elements.size();
+  const int64_t normalized_idx = normalizeIndex(idx, list_size);
+
+  if (normalized_idx < 0 || normalized_idx >= list_size) {
+    if (idx < 0) {
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: This should be normalized_idx right ?

File: torch/csrc/jit/register_prim_ops.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: Yea agreed - If you were implementing GenericList you would probably doing something like ListConstruct where you dispatch to the right function based on the input type of the list.

But aten::eq isn't implemented currently for GenericLists, it's only implemented for our specialized lists. I think it's probably better if you remove GenericList from this PR and leave it for a follow up.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 2 above)
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: This should be normalized_idx right ?

Comment (TheCodez) [reply torch/csrc/jit/register_prim_ops.cpp]: both are fine in this case.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 3]
```diff
@@ -1051,6 +1052,74 @@ int listClear(Stack& stack) {
...
+  } else {
+    AT_ERROR("list.remove(x): x not in list");
+  }
+
+  return 0;
+}
+
+template <>
+int listRemove<Shared<GenericList>, IValue>(Stack& stack) {
+  AT_ERROR("list.remove is not supported yet for 'GenericList'");
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Could you remove this, remove "aten::remove" from the MUTABLE_LIST_OPS macro, and then add add aten::remove with TensorList? 

That way when someone tries to use remove on a generic list the error will be printed out in schema matching, with the source location of the code.

Comment (TheCodez) [reply torch/csrc/jit/register_prim_ops.cpp]: done!

**Review by eellison — APPROVED** [review]
Looks great!! I'll begin landing it now.

**Pull request closed without merging**